### PR TITLE
Fix OP_PUSHDATA encoding

### PR DESCRIPTION
**Pull request by sstone**
There was an error encoding pushes of 255 bytes

### CODE REVIEW

**Review by remyers — CHANGES_REQUESTED** [review]
Seems like the edge cases are handled now. The subtlety with non-minimal encoding tests took me a bit to wrap my head around. 

I have a few questions about the OP_PUSHDATAs < `0x4c` cases and adding exceptions generally. 

I also confirmed the new puzzle by trying with/without the changes you made to `Script.write` and examined the script from the original txs that failed.

File: src/commonMain/kotlin/fr/acinq/bitcoin/Script.kt [hunk 1]
```diff
@@ -95,21 +95,26 @@ public object Script {
             when (head) {
                 is OP_PUSHDATA -> {
                     when {
-                        head.data.size() < 0x4c && head.data.size() == head.code -> {
+                        head.code < 0x4c -> {
+                            require(head.data.size() <= 0x4c)
```
[the remark this answers is not in the document]
Comment (sstone) [reply src/commonMain/kotlin/fr/acinq/bitcoin/Script.kt]: Yes, this check was wrong it should be  ` require(head.data.size() == head.code)`, fixed in  https://github.com/ACINQ/bitcoin-kmp/pull/85/commits/c2c9115babf500de12c1279811dcc05a3bcd2893

File: src/commonMain/kotlin/fr/acinq/bitcoin/Script.kt [hunk 2]
```diff
@@ -95,21 +95,26 @@ public object Script {
...
                             BtcSerializer.writeUInt8(0x4Du, out)
                             BtcSerializer.writeUInt16(head.data.size().toUShort(), out)
                         }
-                        head.data.size() < 0xffffffff && head.code == 0x4e -> {
+                        head.code == 0x4e -> {
+                            require(head.data.size() <= 0xffffffff)
                             BtcSerializer.writeUInt8(0x4Eu, out)
                             BtcSerializer.writeUInt32(head.data.size().toUInt(), out)
                         }
+                        else -> error("invalid OP_PUSHADATA opcode ${head.code}")
```
[the remark this answers is not in the document]
Comment (sstone) [reply src/commonMain/kotlin/fr/acinq/bitcoin/Script.kt:117]: it will be done implicitly now that we check that `head.data.size() == head.code`

File: src/commonTest/kotlin/fr/acinq/bitcoin/BtcSerializerTestsCommon.kt [hunk 3]
```diff
@@ -37,4 +38,31 @@ class BtcSerializerTestsCommon {
...
+        val encoded = Script.write(script)
+        val decoded = Script.parse(encoded)
+        assertEquals(script, decoded)
+
+        fun serde(push: OP_PUSHDATA) {
+            assertEquals(listOf(push), Script.parse(Script.write(listOf(push))))
+        }
+        listOf(
+            OP_PUSHDATA(ByteArray(0x01) { 0x01 }, 0x01),
+            OP_PUSHDATA(ByteArray(0x01) { 0x01 }, 0x4c), // non-minimal encoding
```
[the remark this answers is not in the document]
Comment (sstone) [reply src/commonTest/kotlin/fr/acinq/bitcoin/BtcSerializerTestsCommon.kt:55]: Yes see above

File: src/commonTest/kotlin/fr/acinq/bitcoin/BtcSerializerTestsCommon.kt [hunk 4]
```diff
@@ -37,4 +38,31 @@ class BtcSerializerTestsCommon {
...
+            OP_PUSHDATA(ByteArray(0xffff) { 0x01 }, 0x4e), // non-minimal encoding
+        ).forEach { serde(it) }
+
+        listOf("04deadbeef", "4c04deadbeef", "4d0400deadbeef", "4e04000000deadbeef").forEach { // all encode "deadbeef"
+            val parsed = Script.parse(it)
+            assertEquals(1, parsed.size)
+            assertTrue(ScriptElt.isPush(parsed.first(), 4))
+        }
+    }
 }
```
[the remark this answers is not in the document]
Comment (sstone) [reply src/commonTest/kotlin/fr/acinq/bitcoin/BtcSerializerTestsCommon.kt:78]: It should always be safe to throw an exception here when data is invalid (i.e. does not match the specs), I've added a few additional tests.

**Review by remyers — APPROVED** [review]
LGTM!
